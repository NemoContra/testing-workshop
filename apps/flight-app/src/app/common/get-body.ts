import { inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';

export const getHTMLBodyElement = (): HTMLBodyElement =>
  inject(DOCUMENT).body as HTMLBodyElement;
