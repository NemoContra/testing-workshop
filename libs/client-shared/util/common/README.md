# client-shared-util-common

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test client-shared-util-common` to execute the unit tests.
