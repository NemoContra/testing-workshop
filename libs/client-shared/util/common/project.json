{
  "projectType": "library",
  "sourceRoot": "libs/client-shared/util/common/src",
  "prefix": "testing-workshop",
  "targets": {
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/client-shared/util/common"],
      "options": {
        "jestConfig": "libs/client-shared/util/common/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/client-shared/util/common/**/*.ts",
          "libs/client-shared/util/common/**/*.html"
        ]
      }
    }
  },
  "tags": ["scope:client-shared", "type:util"]
}
