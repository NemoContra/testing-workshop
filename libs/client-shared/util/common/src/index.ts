export * from './lib/pipes/yes-no.pipe';
export * from './lib/helpers/get-query-params';
