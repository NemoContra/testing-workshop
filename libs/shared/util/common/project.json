{
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util/common/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/shared/util/common/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/shared/util/common"],
      "options": {
        "jestConfig": "libs/shared/util/common/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": ["scope:shared", "type:util"]
}
