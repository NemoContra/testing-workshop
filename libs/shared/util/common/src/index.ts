export * from './lib/helpers/create-random-id/create-random-id';
export * from './lib/rxjs-operators/skip-falsy';
