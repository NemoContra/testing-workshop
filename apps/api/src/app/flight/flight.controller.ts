import { Controller, Get, Query } from '@nestjs/common';
import { FlightService } from './flight.service';
import {
  FIND_FLIGHT_FROM,
  FIND_FLIGHT_TO,
  Flight,
} from '@testing-workshop/shared/util/api-interfaces';

@Controller('flight')
export class FlightController {
  constructor(private flightService: FlightService) {}

  @Get()
  getFlights(
    @Query(FIND_FLIGHT_FROM) from?: string,
    @Query(FIND_FLIGHT_TO) to?: string
  ): Flight[] {
    return this.flightService.getFlights({ from, to });
  }
}
