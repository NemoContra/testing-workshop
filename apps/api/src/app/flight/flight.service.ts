import { Injectable } from '@nestjs/common';
import {
  SearchFlightsQueryParams,
  Flight,
} from '@testing-workshop/shared/util/api-interfaces';
import { flights } from './data/flights';

const findOptional = (item: string, optional?: string): boolean =>
  !optional || item.toLowerCase().includes(optional.toLowerCase());

@Injectable()
export class FlightService {
  getFlights({ from, to }: Partial<SearchFlightsQueryParams>): Flight[] {
    return flights
      .filter((flight: Flight) => findOptional(flight.from, from))
      .filter((flight: Flight) => findOptional(flight.to, to));
  }
}
