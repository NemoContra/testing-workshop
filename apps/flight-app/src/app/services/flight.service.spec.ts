import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';
import { FlightService } from './flight.service';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { createSpyObserver } from '@testing-workshop/shared/util/test-helpers';
import { Flight } from '@testing-workshop/shared/util/api-interfaces';
import { HttpStatusCode } from '@angular/common/http';
import { flights } from '@testing-workshop/shared/util/mock-data';

describe('FlightService', () => {
  let spectator: SpectatorService<FlightService>;
  let httpTestingController: HttpTestingController;

  const createService = createServiceFactory({
    service: FlightService,
    imports: [HttpClientTestingModule],
  });

  beforeEach(() => {
    spectator = createService();
    httpTestingController = spectator.inject(HttpTestingController);
  });

  afterEach(() => httpTestingController.verify());

  describe('getFlights', () => {
    it('should getFlights with success', () => {
      const spyObserver = createSpyObserver();
      spectator.service.getFlights().subscribe(spyObserver);

      const testRequest = httpTestingController.expectOne('/api/flight');
      testRequest.flush(flights);

      expect(spyObserver).toMatchSnapshot();
    });

    it('should getFlights with error', () => {
      const spyObserver = createSpyObserver();
      spectator.service.getFlights().subscribe(spyObserver);

      const testRequest = httpTestingController.expectOne('/api/flight');
      testRequest.flush(
        { error: 'Internal Server Error' },
        {
          status: HttpStatusCode.InternalServerError,
          statusText: 'Internal Server Error',
        }
      );

      expect(spyObserver).toMatchSnapshot();
    });
  });

  describe('searchFlights', () => {
    describe.each`
      from         | to           | expectedUrl
      ${undefined} | ${undefined} | ${'/api/flight'}
      ${'Graz'}    | ${undefined} | ${'/api/flight?from=Graz'}
      ${undefined} | ${'Hamburg'} | ${'/api/flight?to=Hamburg'}
      ${'Graz'}    | ${'Hamburg'} | ${'/api/flight?from=Graz&to=Hamburg'}
    `('from $from to $to', ({ from, to, expectedUrl }) => {
      it('should be successful', () => {
        const spyObserver = createSpyObserver();
        spectator.service.searchFlights({ from, to }).subscribe(spyObserver);

        const testRequest = httpTestingController.expectOne(expectedUrl);

        const flight: Flight = {
          id: 1,
          from,
          to,
          date: '2020-10-12T06:22:52.4208588+00:00',
          delayed: false,
        };
        testRequest.flush([flight]);

        expect(spyObserver).toMatchSnapshot();
      });

      it('should fail', () => {
        const spyObserver = createSpyObserver();
        spectator.service.searchFlights({ from, to }).subscribe(spyObserver);

        const testRequest = httpTestingController.expectOne(expectedUrl);

        testRequest.flush(
          { error: 'Internal Server Error' },
          {
            status: HttpStatusCode.InternalServerError,
            statusText: 'Internal Server Error',
          }
        );

        expect(spyObserver).toMatchSnapshot();
      });
    });
  });
});
