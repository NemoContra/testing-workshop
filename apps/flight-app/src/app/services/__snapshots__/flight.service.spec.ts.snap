// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightService getFlights should getFlights with error 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "error": "Internal Server Error",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for /api/flight: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "/api/flight",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`FlightService getFlights should getFlights with success 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Array [
          Object {
            "date": "2020-10-15T06:22:52.4208588+00:00",
            "delayed": false,
            "from": "Graz",
            "id": 1,
            "to": "Mallorca",
          },
          Object {
            "date": "2020-10-12T06:22:52.4208588+00:00",
            "delayed": false,
            "from": "Graz",
            "id": 2,
            "to": "Hamburg",
          },
        ],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FlightService searchFlights from Graz to Hamburg should be successful 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Array [
          Object {
            "date": "2020-10-12T06:22:52.4208588+00:00",
            "delayed": false,
            "from": "Graz",
            "id": 1,
            "to": "Hamburg",
          },
        ],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FlightService searchFlights from Graz to Hamburg should fail 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "error": "Internal Server Error",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for /api/flight?from=Graz&to=Hamburg: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "/api/flight?from=Graz&to=Hamburg",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`FlightService searchFlights from Graz to undefined should be successful 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Array [
          Object {
            "date": "2020-10-12T06:22:52.4208588+00:00",
            "delayed": false,
            "from": "Graz",
            "id": 1,
            "to": undefined,
          },
        ],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FlightService searchFlights from Graz to undefined should fail 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "error": "Internal Server Error",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for /api/flight?from=Graz: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "/api/flight?from=Graz",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`FlightService searchFlights from undefined to Hamburg should be successful 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Array [
          Object {
            "date": "2020-10-12T06:22:52.4208588+00:00",
            "delayed": false,
            "from": undefined,
            "id": 1,
            "to": "Hamburg",
          },
        ],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FlightService searchFlights from undefined to Hamburg should fail 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "error": "Internal Server Error",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for /api/flight?to=Hamburg: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "/api/flight?to=Hamburg",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;

exports[`FlightService searchFlights from undefined to undefined should be successful 1`] = `
Object {
  "complete": [MockFunction complete] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "error": [MockFunction error],
  "next": [MockFunction next] {
    "calls": Array [
      Array [
        Array [
          Object {
            "date": "2020-10-12T06:22:52.4208588+00:00",
            "delayed": false,
            "from": undefined,
            "id": 1,
            "to": undefined,
          },
        ],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;

exports[`FlightService searchFlights from undefined to undefined should fail 1`] = `
Object {
  "complete": [MockFunction complete],
  "error": [MockFunction error] {
    "calls": Array [
      Array [
        HttpErrorResponse {
          "error": Object {
            "error": "Internal Server Error",
          },
          "headers": HttpHeaders {
            "headers": Map {},
            "lazyUpdate": null,
            "normalizedNames": Map {},
          },
          "message": "Http failure response for /api/flight: 500 Internal Server Error",
          "name": "HttpErrorResponse",
          "ok": false,
          "status": 500,
          "statusText": "Internal Server Error",
          "url": "/api/flight",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "next": [MockFunction next],
}
`;
